src/ucp/rndv/proto_rndv.c

```diff
@@ -175,26 +178,39 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params)
     }
 
     /* Set rendezvous protocol properties */
-    *params->super.super.priv_size         = sizeof(ucp_proto_rndv_ctrl_priv_t);
-    params->super.super.caps->cfg_thresh   = params->super.cfg_thresh;
-    params->super.super.caps->cfg_priority = params->super.cfg_priority;
-    params->super.super.caps->min_length   = params->min_length;
-    params->super.super.caps->num_ranges   = 0;
+    ucp_proto_select_get_valid_range(rpriv->remote_proto.thresholds,
+                                     &caps->min_length, &max_length);
+    caps->cfg_thresh   = params->super.cfg_thresh;
+    caps->cfg_priority = params->super.cfg_priority;
+    caps->num_ranges   = 0;
 
     /* Copy performance ranges from the remote protocol, and add overheads */
-    remote_perf_range = rpriv->remote_proto.perf_ranges;
-    caps              = params->super.super.caps;
+    remote_range = rpriv->remote_proto.perf_ranges;
+    rndv_bias    = ucs_linear_func_make(0, 1.0 - params->perf_bias);
     do {
         perf_range             = &caps->ranges[caps->num_ranges];
-        perf_range->max_length = remote_perf_range->max_length;
+        perf_range->max_length = remote_range->super.max_length;
+        if (perf_range->max_length < caps->min_length) {
+            continue;
+        }
 
-        /* Add send overheads and apply perf_bias */
-        perf_range->perf = ucs_linear_func_compose(
-                ucs_linear_func_make(0, 1.0 - params->perf_bias),
-                ucs_linear_func_add(remote_perf_range->perf, send_overheads));
+        /* Single */
+        perf = ucs_linear_func_add(
+                send_overheads,
+                remote_range->super.perf[UCP_PROTO_PERF_TYPE_SINGLE]);
+        perf_range->perf[UCP_PROTO_PERF_TYPE_SINGLE] =
+                ucs_linear_func_compose(rndv_bias, perf);
+
+        /* Pipelined */
+        perf = ucp_proto_common_ppln_perf(
+                send_overheads,
+                remote_range->super.perf[UCP_PROTO_PERF_TYPE_MULTI],
+                perf_range->max_length);
+        perf_range->perf[UCP_PROTO_PERF_TYPE_MULTI] =
+                ucs_linear_func_compose(rndv_bias, perf);
```

Comment:
why do we need pipeline for ctrl messages?

Reply:
because when calculating the performance of 3 stage pipeline, where each fragment sends ATP, the pipeline mode should not count the ack latency

Reply:
I guess it is for compliance, since RNDV RTS can be 128 bytes, but AM Bcopy max is 64 bytes